src/snowflake/snowpark/_internal/temp_table_cleaner.py

```diff
@@ -0,0 +1,78 @@
+#
+
+import logging
+import threading
+from collections import defaultdict
+from queue import Queue
+from typing import TYPE_CHECKING, Dict
+
+if TYPE_CHECKING:
+    from snowflake.snowpark.session import Session  # pragma: no cover
+
+
+class TempTableCleaner:
+    """
+    Clean up temporary tables created automatically by Snowpark at the session level.
+
+    This function identifies and removes any temporary tables that have been
+    generated by Snowpark during the current session. Temporary tables are
+    typically used for intermediate computations (e.g., df.cache_result) and
+    are not needed when they are not referenced anymore. Removing these tables
+    helps maintain a clean working environment and reduce storage cost for a long-running
+    session.
+    """
+
+    def __init__(self, session: "Session") -> None:
+        self.session = session
+        # this dict maintains key-value pair from Snowpark-generated temp table fully-qualified name
+        # to its reference count for later temp table management
+        # this dict will still be maintained even if the cleaner is stopped (`stop()` is called)
+        self.ref_count_map: Dict[str, int] = defaultdict(int)
+        self.queue = Queue()
+        self.cleanup_thread = None
+
+    def inc_ref_count(self, name: str) -> None:
+        self.ref_count_map[name] += 1
+
+    def dec_ref_count(self, name: str) -> None:
+        self.ref_count_map[name] -= 1
+        if self.ref_count_map[name] <= 0:
+            self.ref_count_map.pop(name)
+            # clean up
+            if self.is_alive():
+                self.queue.put(name)
+
+    def process_cleanup(self) -> None:
+        while True:
+            table_name = self.queue.get()
+            if table_name is None:
+                # exit the thread loop
+                break
+            self.drop_table(table_name)
+
+    def drop_table(self, name: str) -> None:
+        common_log_text = f"temp table {name} in session {self.session.session_id}"
+        try:
+            # TODO SNOW-1556553: Remove this workaround once multi-threading of Snowpark session is supported
+            with self.session._conn._conn.cursor() as cursor:
+                cursor.execute(f"drop table if exists {name}")
+            logging.debug(f"Cleanup Thread: Successfully dropped {common_log_text}")
+        except Exception:
+            logging.debug(
+                f"Cleanup Thread: Failed to drop {common_log_text}"
+            )  # pragma: no cover
+
+    def is_alive(self) -> bool:
+        return self.cleanup_thread is not None and self.cleanup_thread.is_alive()
+
+    def start(self) -> None:
+        if not self.is_alive():
+            self.cleanup_thread = threading.Thread(target=self.process_cleanup)
+            self.cleanup_thread.start()
+
+    def stop(self) -> None:
+        if self.is_alive():
+            self.queue.put(None)
+            self.cleanup_thread.join()
```

Comment:
let's also clean up the ref_count_map

Reply:
That makes me start thinking of a scenario where if user start turning this process on and off, which may not working well.
for example, 
1) one ref of table1 is established
2) clean up turned off
3) an other table1 ref is established
4) turn clean up on, and one table1 ref is dropped
we might drop off the table1 incorrectly.

in python is there a way to discover all SnowflakeTable object to recover the ref state correctly? that is similar as service recover, where we need to recover the service state correctly when the service is back on, otherwise the service will not operate correctly.